sklearn/metrics/pairwise.py

```diff
@@ -1195,12 +1196,19 @@ def pairwise_distances(X, Y=None, metric="euclidean", n_jobs=1, **kwds):
                             " support sparse matrices.")
         X, Y = check_pairwise_arrays(X, Y)
         if n_jobs == 1 and X is Y:
+            if metric in PAIRWISE_BOOLEAN_FUNCTIONS and X.dtype != bool:
+                warnings.warn("Metric %s: Implicit cast to boolean for X of "
+                              "dtype %s" % (metric, X.dtype))
+                X = X.astype(np.bool)
             return distance.squareform(distance.pdist(X, metric=metric,
                                                       **kwds))
         func = partial(distance.cdist, metric=metric, **kwds)
 
     return _parallel_pairwise(X, Y, func, n_jobs, **kwds)
 
+PAIRWISE_BOOLEAN_FUNCTIONS = [
+    'jaccard',
```

Comment:
there are others like ``yule`` mentioned in #5333. But I guess we haven't implemented them. So this is the only one that the trees support?
Can you add a non-regression test that checks that nearest neighbors with a tree gives the same result as brute on float input? and check that if fails on master?